### ISSUE TITLE
bundle exec ruby ignores arguments like --jvm

### ISSUE THREAD

**Issue by chrisseaton** [oracle/truffleruby#2108]
```
% bundle exec ruby --jvm --version            
truffleruby (Shopify) 20.3.0-dev-c9a18fad, like ruby 2.6.6, GraalVM CE Native [x86_64-darwin]
% ruby --jvm --version
truffleruby (Shopify) 20.3.0-dev-c9a18fad, like ruby 2.6.6, GraalVM CE JVM [x86_64-darwin]
 % TRUFFLERUBYOPT=--jvm bundle exec ruby --version  
truffleruby (Shopify) 20.3.0-dev-c9a18fad, like ruby 2.6.6, GraalVM CE JVM [x86_64-darwin]
```

**Comment by eregon** [comment]
I can reproduce.
It seems to work on MRI, so might be a TruffleRuby bug:
```
$ bundle exec ruby --jit --version    
ruby 2.6.6p146 (2020-03-31 revision 67876) +JIT [x86_64-linux]

$ bundle exec ruby --jvm --version
truffleruby 20.2.0, like ruby 2.6.6, GraalVM CE Native [x86_64-linux]
```

If I use `export VERBOSE_GRAALVM_LAUNCHERS=true`, and change `bin/bundle` to use:
```bash
echo "$(dirname $SELF_PATH)/ruby" "$SELF_PATH" "$@"
"$(dirname $SELF_PATH)/ruby" -v
echo
exec "$(dirname $SELF_PATH)/ruby" "$SELF_PATH" "$@"
```
we get
```
$ bundle exec ruby -w --jvm -w --version
/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/bin/ruby /home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/bin/bundle exec ruby -w --jvm -w --version
truffleruby 20.2.0, like ruby 2.6.6, GraalVM CE Native [x86_64-linux]

exec: /home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/bin/java -classpath /home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/truffleruby-annotations.jar:/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/truffleruby-shared.jar:/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/lib/graalvm/launcher-common.jar:/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/lib/graalvm/truffleruby-launcher.jar org.truffleruby.launcher.RubyLauncher --disable-rubyopt /home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/bin/bundle exec ruby -w -w --version
truffleruby 20.2.0, like ruby 2.6.6, GraalVM CE Native [x86_64-linux]
```

So first, `bundle` will run TruffleRuby Native, with those arguments, and not see the `--jvm` option, as expected (it's an application argument at that point).

Then at some point Bundler will `execve()` to execute `ruby`.
At that point `--jvm` seems still there, because the native launcher sees somehow that it needs to switch to JVM and does that (the command is `java ...`).
In the process the native launcher seems to remove the --jvm argument.

Still a puzzle why the `--version` prints the native one.

**Comment by eregon** [comment]
Actually it seems we look into application arguments in the native launcher, because:
```ruby
# foo.rb
p [:ARGV, ARGV]
puts RUBY_DESCRIPTION
```

```
$ ruby foo.rb --jvm -e 'p RUBY_DESCRIPTION'
exec: /home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/bin/java -classpath /home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/truffleruby-annotations.jar:/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/languages/ruby/truffleruby-shared.jar:/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/lib/graalvm/launcher-common.jar:/home/eregon/.rubies/truffleruby+graalvm-20.2.0/graalvm/jre/lib/graalvm/truffleruby-launcher.jar org.truffleruby.launcher.RubyLauncher --disable-rubyopt foo.rb -e 'p RUBY_DESCRIPTION'
truffleruby 20.2.0, like ruby 2.6.6, GraalVM CE JVM [x86_64-linux]
```
But that should pass `--jvm` as an application argument and not switch to JVM.

**Comment by eregon** [comment]
As a note, the Bash launchers used in developments (`mx.truffleruby/jvm*`) have no chance to know if `--jvm` is in application arguments or not (or they would need to launch twice the JVM to know), so that's a known limitation for Bash launchers.

But the native launcher should behave correctly because it does parse the command line (twice when exec'ing to JVM).

**Comment by eregon** [comment]
This is an issue of GraalVM launchers in general, I have a patch locally, I'll try to integrate it.

**Comment by eregon** [comment]
Fixed by https://github.com/oracle/graal/commit/8a96bec03e19a666ae749cce35f8e895a6561844 (fixes in SDK Launcher) and b9a4cbdc9351b012bd82848291c6544ec4d87515 (specs for it & simplifications).

**Comment by chrisseaton** [comment]
Great thanks very much for that.
